### PR TITLE
[SDK Updates] Kotlin migration

### PR DESCRIPTION
**Pull request by mpetrenco**
# Changelog:
- Migrated from Java to Kotlin;
- Added type safety for all property mappings that could throw incorrect type exceptions;
- Added mapping for PrimaryAccountDetails;

### CODE REVIEW

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 1]
```diff
@@ -0,0 +1,651 @@
...
+import java.lang.Exception
+import kotlin.collections.ArrayList
+
+class JudoReactNativeModule internal constructor(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {
+
+    // ------------------------------------------------------------
+    // MARK: Constants
+    // ------------------------------------------------------------
+
+    companion object {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:25]: No need for this, just declare it as a constant outside this class
`const val JUDO_PAYMENT_WIDGET_REQUEST_CODE = 1`

File: types/JudoTypes.tsx [hunk 2]
```diff
@@ -35,8 +35,8 @@ export interface JudoAmount {
 
 export interface JudoReference {
     consumerReference: string,
-    paymentReference: string,
-    metadata?: Map<String, String>,
+    paymentReference?: string,
+    metadata?: Record<string, string>,
```
Comment (stpopa) [inline types/JudoTypes.tsx:39]: `metadata?: Record<string, any>` ?

File: android/src/main/java/com/reactlibrary/JudoReactNativePackage.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import com.facebook.react.bridge.ReactApplicationContext
+import com.facebook.react.uimanager.ViewManager
+import java.util.*
+
+@Suppress("unused")
+class JudoReactNativePackage : ReactPackage {
+    override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> {
+        val modules: MutableList<NativeModule> = ArrayList()
+        modules.add(JudoReactNativeModule(reactContext))
+        return modules
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativePackage.kt:14]: you can just: `return arrayOf(JudoReactNativeModule(reactContext))`

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+        private const val JUDO_PAYMENT_WIDGET_REQUEST_CODE = 1
+    }
+
+    // ------------------------------------------------------------
+    // MARK: Variables
+    // ------------------------------------------------------------
+
+    private var transactionPromise: Promise? = null
+
+    private val listener: ActivityEventListener = object : BaseActivityEventListener() {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:35]: you can `private val listener = object : BaseActivityEventListener() { `

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+
+        override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent) {
+
+            if (resultCode == PAYMENT_ERROR) {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:39]: can we replace these if's with a `when` statement? https://kotlinlang.org/docs/reference/control-flow.html#when-expression

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 6]
```diff
@@ -0,0 +1,669 @@
...
+                val error: ApiError? = data.getParcelableExtra(JUDO_ERROR)
+                error?.let {
+                    transactionPromise?.reject(error.code.toString(), error.message)
+                }
+            }
+
+            if (resultCode == PAYMENT_SUCCESS) {
+                val receipt: Receipt? = data.getParcelableExtra(JUDO_RECEIPT)
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:47]: `val receipt = data.getParcelableExtra<Receipt>(JUDO_RECEIPT) `

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+    // ------------------------------------------------------------
+
+    private var transactionPromise: Promise? = null
+
+    private val listener: ActivityEventListener = object : BaseActivityEventListener() {
+
+        override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent) {
+
+            if (resultCode == PAYMENT_ERROR) {
+                val error: ApiError? = data.getParcelableExtra(JUDO_ERROR)
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:40]: `val error = data.getParcelableExtra<ApiError>(JUDO_ERROR)`

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+
+    // ------------------------------------------------------------
+    // MARK: Variables
+    // ------------------------------------------------------------
+
+    private var transactionPromise: Promise? = null
+
+    private val listener: ActivityEventListener = object : BaseActivityEventListener() {
+
+        override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent) {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:37]: please add a check for the `requestCode == JUDO_PAYMENT_WIDGET_REQUEST_CODE` to filter out the rest of `onActivityResult`s

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+
+        val type = when (transactionMode) {
+            0 -> PaymentWidgetType.PAYMENT_METHODS
+            else -> PaymentWidgetType.PRE_AUTH_PAYMENT_METHODS
+        }
+
+        return getJudoConfiguration(type, options)
+    }
+
+    private fun getJudoConfiguration(type: PaymentWidgetType, options: ReadableMap): Judo? {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:139]: can we avoid optionals here,
let's return a `Judo` object, and re-throw the exception to be handled by the caller:

```
    fun invokeGooglePay(options: ReadableMap, promise: Promise) {
        try {
        	val judo = getGoogleTransactionConfiguration(options)
			startJudoActivity(judo, promise)
        } catch (error: Error) {
        	promise.reject(error)
        }
    }

```

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 10]
```diff
@@ -0,0 +1,669 @@
...
+        }
+
+        return try {
+            builder.build()
+        } catch (_: Exception) {
+            null
+        }
+    }
+
+    private fun getWidgetType(options: ReadableMap): PaymentWidgetType {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:207]: you can simplify this stuff:


   ```
 private fun getWidgetType(options: ReadableMap) =  when (options.getInt("transactionType")) {
            1 -> PaymentWidgetType.PRE_AUTH_CARD_PAYMENT
            2 -> PaymentWidgetType.CREATE_CARD_TOKEN
            3 -> PaymentWidgetType.CHECK_CARD
            4 -> PaymentWidgetType.SAVE_CARD
            else -> PaymentWidgetType.CARD_PAYMENT
        }
```

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt [hunk 11]
```diff
@@ -0,0 +1,669 @@
...
+        }
+
+        return try {
+            builder.build()
+        } catch (_: Exception) {
+            null
+        }
+    }
+
+    private fun getSiteId(options: ReadableMap): String? {
```
Comment (stpopa) [inline android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:322]: to make this file shorter, let's extract this stuff to extension functions, ex:
```
val Any.configuration: Map<String, Any>?
    get() = try {
        getMap("configuration")
    } catch (_: Exception) {
        null
    }

val Any.siteId: String?
    get() = try {
    configuration?.getString("siteId")
} catch (_: Exception) {
    null
}

```

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 1 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativePackage.kt (hunk 3 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativePackage.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 4 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 5 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 7 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 6 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 8 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:32]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 9 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt:133]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 10 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt (hunk 11 above)
Comment (mpetrenco) [reply android/src/main/java/com/reactlibrary/JudoReactNativeModule.kt]: Fixed

File: types/JudoTypes.tsx (hunk 2 above)
Comment (mpetrenco) [reply types/JudoTypes.tsx:39]: Still have no idea how to add dynamic values to a Bundle.

Comment (stpopa) [reply types/JudoTypes.tsx:39]: ok, I'll have a look later this week

**Pull request merged**